src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -249,70 +205,135 @@ protected virtual string GenerateUniqueId()
+                    if (!renderedBoundAttributeNames.Add(attribute.Key))
                     {
+                        // We only need to create attribute values once per HTML element (not once per tag helper).
                         continue;
                     }
 
-                    // We need to inform the context of the attribute value.
-                    _writer
-                        .WriteStartInstanceMethodInvocation(
-                            ExecutionContextVariableName,
-                            _tagHelperContext.ExecutionContextAddTagHelperAttributeMethodName)
-                        .WriteStringLiteral(attributeName)
-                        .WriteParameterSeparator()
-                        .Write(valueAccessor)
-                        .WriteEndMethodInvocation();
+                    // We need to capture the tag helper's property value accessor so we can retrieve it later
+                    // (if we need to).
+                    string valueAccessor = null;
+
+                    foreach (var associatedDescriptor in associatedDescriptors)
+                    {
+                        var associatedAttributeDescriptor = associatedDescriptor.Attributes.First(
+                            attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Key));
+                        var tagHelperVariableName = GetVariableName(associatedDescriptor);
+
+                        valueAccessor = RenderBoundAttribute(
+                            tagHelperVariableName,
+                            valueAccessor,
+                            attribute,
+                            associatedAttributeDescriptor);
+                    }
                 }
                 else
                 {
-                    // The attribute value has already been recorded, lets retrieve it from the stored value
-                    // accessors.
-                    _writer
-                        .WriteStartAssignment(valueAccessor)
-                        .Write(htmlAttributeValues[attributeName])
-                        .WriteLine(";");
+                    RenderUnboundAttribute(attribute);
+                }
+            }
+        }
+
+        private string RenderBoundAttribute(
+            string tagHelperVariableName,
+            string previousValueAccessor,
+            KeyValuePair<string, Chunk> attribute,
+            TagHelperAttributeDescriptor attributeDescriptor)
+        {
+            var attributeName = attribute.Key;
+            var attributeValueChunk = attribute.Value;
+            if (attributeValueChunk == null)
+            {
+                // Minimized attributes are not valid for bound attributes. TagHelperBlockRewriter has already
+                // logged an error if it was a bound attribute; so we can skip.
+                return null;
+            }
+
+            // We need to capture the tag helper's property value accessor so we can retrieve it later (if we need to).
```

Comment:
Alternatively use `// Haven't seen this attribute value before and need to record its value.`